### ISSUE TITLE
spacePerIndentLevel value not completely taken in account

### ISSUE THREAD

**Issue by nvuillam** [CodeNarc/CodeNarc#553]
According to the documentation, when we want to override the indentation level , we can define a value for spacePerIndentLevel property

If we define 2, we would expect the first level starting at 0, the second level at 2, the thirst a 4, etc ...

But when we lint a 4-indent-level formatted file with spacePerIndentationLevel=2, no violation is detected

How to reproduce :
- Clone CodeNarc repo
- Add method _test_Method_spacesPerIndentLevelOverride_Violation_ in class IndentationRuleTest.groovy
- Follow examples below 

### Test 1: spacesPerIndentLevel = 2 
- 0 violation found
- **Wrong behavior**
- Should be found because `description('asd')` is 4 chars away from `listView('123')`, whereas one level is only 2 spaces

```groovy
    @Test
    void test_Method_spacesPerIndentLevelOverride_Violation() {
        final SOURCE = '''
            |listView('123') {
            |    description('asd')
            |    jobs {
            |        name('qwe')
            |        name('zxc')
            |    }
            |    columns {
            |        status()
            |        weather()
            |    }
            |}
        '''.stripMargin()
        rule.spacesPerIndentLevel = 2
        assertViolations(SOURCE,
                [lineNumber:3],
                [lineNumber:4],
                [lineNumber:5],
                [lineNumber:6],
                [lineNumber:8],
                [lineNumber:9],
                [lineNumber:10]
        )
    }
```

### Test 2: spacesPerIndentLevel = 3
- 7 violations found 
- **correct behaviour**

```groovy
    @Test
    void test_Method_spacesPerIndentLevelOverride_Violation() {
        final SOURCE = '''
            |listView('123') {
            |    description('asd')
            |    jobs {
            |        name('qwe')
            |        name('zxc')
            |    }
            |    columns {
            |        status()
            |        weather()
            |    }
            |}
        '''.stripMargin()
        rule.spacesPerIndentLevel = 3
        assertViolations(SOURCE,
                [lineNumber:3],
                [lineNumber:4],
                [lineNumber:5],
                [lineNumber:6],
                [lineNumber:8],
                [lineNumber:9],
                [lineNumber:10]
        )
    }
```

Test 2: spacesPerIndentLevel = 4
- 0 violation found 
- **correct behavior**

```groovy
    @Test
    void test_Method_spacesPerIndentLevelOverride_Violation() {
        final SOURCE = '''
            |listView('123') {
            |    description('asd')
            |    jobs {
            |        name('qwe')
            |        name('zxc')
            |    }
            |    columns {
            |        status()
            |        weather()
            |    }
            |}
        '''.stripMargin()
        rule.spacesPerIndentLevel = 4
        assertViolations(SOURCE,
                [lineNumber:3],
                [lineNumber:4],
                [lineNumber:5],
                [lineNumber:6],
                [lineNumber:8],
                [lineNumber:9],
                [lineNumber:10]
        )
    }
```

- As npm-groovy-lint often uses CodeNarc responses to apply format & fixes, I'm currently in no position to offer users to decide to use 2 spaces by indentation level 😢 

- Related npm-groovy-lint issue : [#99](https://github.com/nvuillam/npm-groovy-lint/issues/99)

**Comment by chrismair** [comment]
The *Indentation* rule is easily the most troublesome and challenging rule in CodeNarc. There are just so many styles and variations and nooks and crannies in the AST and language changes, etc..

Actually that rule has gotten more *lenient*, rather than more *strict*, in order to reduce brittleness. So the *spacesPerIndentLevel* property is used in multiple places to constrain the allowed indent columns, rather than to enforce exact values. So, if it is set to 4, then the allowed indent columns may be 13, 17 or 21 (1-based, rather than 0-based) for a particular statement.

**Comment by nvuillam** [comment]
I always use the first value of such list :)

Would it be complicated to have a parameter like "lenient" true/false with true by default , so when I use lenient=false I have only one possible indentation value ?

**Comment by chrismair** [comment]
Yes, unfortunately that would make the rule more brittle and wrong more often. Part of the challenge is because of all of the styles/constructs/variations/AST, it is difficult to always know the exact right column.

**Comment by nvuillam** [comment]
It worked very well will the previous version, if i create branches
between  previous and new behavior would it be ok ?

Le ven. 11 sept. 2020 à 00:48, Chris Mair <notifications@github.com> a
écrit :

**Comment by chrismair** [comment]
I'm not sure what you are asking.

**Comment by nvuillam** [comment]
Basically, have some switch in the code, based on RuleProperty lenient true/false :
- 1: IndentationRule behaviour since 1.6 : Several possibilities for Indentation ( ex: columns [3,5,7]
- 2: IndentationRule behaviour before 1.6: Only one possibility for Indentation ( ex: 3 )

**Comment by chrismair** [comment]
I don't think that would be sustainable. The reason the leniency was added was to make the rule more resilient, and adding that "strict" option would undo that.

**Comment by nvuillam** [comment]
I get your point... but I'm kind of stucked :/ 
How can I reformat with correct indentation (2, defined by user) , if CodeNarc does not reply the info that the indentation is wrong ? :/

**Comment by tthornton3-chwy** [comment]
@chrismair I understand the concern. We don't want to add "gotchas" and (forever having to support) "one-offs" into a codebase for something that seems niche. I think that's completely reasonable! However in this case, this isn't niche, I was able to find a good amount of people reporting on this being an issue on one of many wrappers around this amazing tool, but not necessarily making it back to here. I have also bumped into this. We would love a way to Indent properly, and currently/unfortunately the new behavior won't support that.

What @nvuillam is proposing does not inherently undo your efforts of making the rule more resilient in itself. It would be opt-in, providing the exact experience one would expect by default :) And just in the cases where this pops up, like for any of the wrappers that touch Jenkinsfiles, it can be enabled. Would it help you feel better if they/I did the work to add this, making sure to have a very good amount of tests?

**Comment by chrismair** [comment]
I am not opposed to providing support for that. Further discussion and PRs are welcome.

My concerns are mostly around maintainability.

If we introduced a "lenient" flag and conditional behavior, it would need to:
 - Not affect (break) existing users
 - Not make the existing logic unmaintainable (and like I said above, it is already dangerously close to that)
 - Probably start making more "opinionated" choices about what is the "right" way to format. That is certainly one of the challenges for this rule is the fact that a lot of decisions come down to preference/convention rather than right versus wrong -- there are often multiple ways to reasonably indent things.

And of course, the initial development of new features and enhancements like this is just a (small) part of the work involved -- that maintaining it over the next 10+ years can become unpleasant if we make the wrong choices.  That should not prevent forward progress, though, and I am not opposed to exploring making this rule better and more useful.
